### PR TITLE
Improve fabric_util get_db timeout logic

### PR DESCRIPTION
**Pull request by nickva**
Previously, users with low {Q, N} dbs often got the `"No DB shards could be opened."` error when the cluster is overloaded. The hard-coded 100 msec timeout was too low to open the few available shards and the whole request would crash with a 500 error.

Attempt to calculate an optimal timeout value based on the number of shards and the max fabric request timeout limit.

The sequence of doubling (by default) timeouts forms a geometric progression. Use the well known closed form formula for the sum [0], and the maximum request timeout, to calculate the initial timeout. The test case illustrates a few examples with some default Q and N values.

Because we don't want the timeout value to be too low, since it takes time to open shards, and we don't want to quickly cycle through a few initial shards and discard the results, the minimum initial timeout is clipped to the
previously hard-coded 100 msec timeout. Unlike previously however, this minimum value can now also be configured.

[0] https://en.wikipedia.org/wiki/Geometric_series

Fixes: https://github.com/apache/couchdb/issues/3733

### CODE REVIEW

File: src/fabric/src/fabric_util.erl [hunk 1]
```diff
@@ -100,19 +100,22 @@ get_db(DbName) ->
     get_db(DbName, []).
 
 get_db(DbName, Options) ->
-    {Local, SameZone, DifferentZone} = mem3:group_by_proximity(mem3:shards(DbName)),
+    LiveShards = mem3:live_shards(DbName, [node() | nodes()]),
```
Comment (rnewson) [inline src/fabric/src/fabric_util.erl:103]: this change is unrelated to the PR topic.

File: src/fabric/src/fabric_util.erl [hunk 2]
```diff
@@ -125,15 +128,40 @@ get_shard([#shard{node = Node, name = Name} | Rest], Opts, Timeout, Factor) ->
...
+    % Where T is the initial timeout and F is the factor. The formula for
+    % the sum is:
+    %
+    %     Sum[T * F^I, I <- 0..N] = T * (1 - F^(N + 1)) / (1 - F)
+    %
+    % Then, for a given sum and factor we can calculate the initial timeout T:
+    %
+    %     T = Sum / ((1 - F^(N+1)) / (1 - F))
+    %
+    Timeout = MaxTimeout / ((1 - math:pow(Factor, N + 1)) / (1 - Factor)),
```
Comment (rnewson) [inline src/fabric/src/fabric_util.erl:159]: we have exponential backoff algorithms elsewhere (and maybe some simple linear ones too), the proliferation of different timeout algorithms is confusing.

**Review by rnewson — CHANGES_REQUESTED** [review]
the basic idea of allowing the hardcoded `100` initial timeout to change is definitely useful but should be presented in isolation. the other changes can then be debated in other PR's.

File: src/fabric/src/fabric_util.erl (hunk 2 above)
Comment (nickva) [reply src/fabric/src/fabric_util.erl:159]: This is the opposite calculation, given the max time we are willing to wait, and the factor, calculate the initial timeout.

File: src/fabric/src/fabric_util.erl (hunk 1 above)
Comment (nickva) [reply src/fabric/src/fabric_util.erl:103]: fair enough

**Review by rnewson — APPROVED** [review]
